src/Workspaces/Core/Portable/Workspace/Solution/AnalyzerConfigDocumentState.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class AnalyzerConfigDocumentState : TextDocumentState
+    {
+        private readonly ValueSource<AnalyzerConfig> _analyzerConfigValueSource;
+
+        private AnalyzerConfigDocumentState(
+            SolutionServices solutionServices,
+            IDocumentServiceProvider documentServiceProvider,
+            DocumentInfo.DocumentAttributes attributes,
+            SourceText sourceTextOpt,
+            ValueSource<TextAndVersion> textAndVersionSource)
+            : base(solutionServices, documentServiceProvider, attributes, sourceTextOpt, textAndVersionSource)
+        {
+            _analyzerConfigValueSource = CreateAnalyzerConfigValueSource();
+        }
+
+        public AnalyzerConfigDocumentState(
+            DocumentInfo documentInfo,
+            SolutionServices solutionServices)
+            : base(documentInfo, solutionServices)
+        {
+            _analyzerConfigValueSource = CreateAnalyzerConfigValueSource();
+        }
+
+        private ValueSource<AnalyzerConfig> CreateAnalyzerConfigValueSource()
+        {
+            return new AsyncLazy<AnalyzerConfig>(
+                asynchronousComputeFunction: async cancellationToken => AnalyzerConfig.Parse(await GetTextAsync(cancellationToken).ConfigureAwait(false), FilePath),
+                synchronousComputeFunction: cancellationToken => AnalyzerConfig.Parse(GetTextSynchronously(cancellationToken), FilePath),
```

Comment:
fyi, i found this super hard to parse in my head.

Reply:
I thought we used to have some ValueSource that was "given another ValueSource, please apply this transformation onto it" but I couldn't find it. Really the ugliness of this is mostly boilerplate.